addons/purchase/tests/test_create_picking.py

```diff
@@ -89,3 +89,50 @@ def test_01_check_double_validation(self):
         # PO approved by manager
         self.po.button_approve()
         self.assertEqual(self.po.state, 'purchase', 'PO state should be "Purchase".')
+
+    def test_02_po_return(self):
+        """
+        Test a PO with a product on delivery. Deliver the PO, then do a return
+        of the picking with Refund.
+        """
+        # Draft purchase order created
+        self.po = self.env['purchase.order'].create(self.po_vals)
+        self.assertTrue(self.po, 'Purchase: no purchase order created')
+        self.assertEqual(self.po.order_line.mapped('qty_received'), [0.0], 'Purchase: no product should be received"')
+        self.assertEqual(self.po.order_line.mapped('qty_invoiced'), [0.0], 'Purchase: no product should be invoiced"')
+
+        self.po.button_confirm()
+        self.assertEqual(self.po.state, 'purchase', 'Purchase: PO state should be "Purchase"')
+        self.assertEqual(self.po.invoice_status, 'to invoice', 'Purchase: PO invoice_status should be "Waiting Invoices"')
+
+        # Purchase order confirm
+        self.po.button_confirm()
+        self.assertEqual(self.po.state, 'purchase', 'Purchase: PO state should be "Purchase')
+        self.assertEqual(self.po.picking_count, 1, 'Purchase: one picking should be created"')
+        self.picking = self.po.picking_ids[0]
+        self.picking.force_assign()
+        self.picking.pack_operation_product_ids.write({'qty_done': 5.0})
+        self.picking.do_new_transfer()
+        self.assertEqual(self.po.order_line.mapped('qty_received'), [5.0], 'Purchase: all products should be received"')
+
+        # Check quantity delivered
+        del_qty = sum(pol.qty_received for pol in self.po.order_line)
+        self.assertEqual(del_qty, 5.0, 'Purchase: delivered quantity should be 5.0 instead of %s after complete delivery' % del_qty)
+
+        # Create return picking
+        StockReturnPicking = self.env['stock.return.picking']
+        pick = self.po.picking_ids
+        default_data = StockReturnPicking.with_context(active_ids=pick.ids, active_id=pick.ids[0]).default_get(['move_dest_exists', 'original_location_id', 'product_return_moves', 'parent_location_id', 'location_id'])
+        return_wiz = StockReturnPicking.with_context(active_ids=pick.ids, active_id=pick.ids[0]).create(default_data)
+        return_wiz.product_return_moves.quantity = 2.0 # Return only 2
+        return_wiz.product_return_moves.to_refund = True # Refund these 2
+        res = return_wiz.create_returns()
+        return_pick = self.env['stock.picking'].browse(res['res_id'])
+
+        # Validate picking
+        return_pick.force_assign()
+        return_pick.pack_operation_product_ids.write({'qty_done': 2})
+        return_pick.do_new_transfer()
+
+        # Check delivered quantity
+        self.assertEqual(self.po.order_line[0].qty_received, 3.0, 'Purchase: delivered quantity should be 3.0 instead of "%s" after picking return' % self.po.order_line[0].qty_received)
```

Comment:
It's nice to see tests added with new functionalities :-) 

(though it doesn't look like it tests the created invoices :p )